distribution/lib/Standard/Base/0.0.0-dev/src/Data/Statistics.enso

```diff
@@ -139,44 +153,70 @@ type Statistic
-                _ -> stats_array.at i
+        counter = CountMinMax.new
+
+        java_stats = statistics.map .order
+        skip_moments = java_stats.all s->s.is_nothing
+        moments = if skip_moments then Nothing else Moments.new (java_stats.filter v->v!=Nothing . fold 0 .max)
+
+        result = data.fold [counter, moments] compute_fold
+
+        report_error stats =
+            first = 0.up_to stats.length . find i-> (java_stats.at i . is_nothing . not)
+            stat = stats.at first
+            Error.throw (Illegal_Argument.Error ("Can only compute " + stat.to_text + " on numerical data sets."))
+
+        if counter.count == 0 then statistics.map empty_value else
+            if skip_moments.not && result.second.is_nothing then report_error statistics else
+                statistics.map s-> case s of
+                    Statistic.Count -> counter.count
+                    Statistic.Minimum -> if counter.comparatorError then (Error.throw Incomparable_Values) else counter.minimum
+                    Statistic.Maximum -> if counter.comparatorError then (Error.throw Incomparable_Values) else counter.maximum
+                    Statistic.Covariance s -> calculate_correlation_statistics data s . covariance
+                    Statistic.Pearson s -> calculate_correlation_statistics data s . pearsonCorrelation
+                    Statistic.Spearman s -> calculate_spearman_rank data s
+                    Statistic.R_Squared s -> calculate_correlation_statistics data s . rSquared
+                    _ -> moments.compute (to_moment_statistic s)
+
+    ## Compute a running statistics on a vector like object.
+
+       Arguments:
+       - data: Vector like object which has a `to_array` method.
+       - statistics: Set of statistics to calculate.
+    running : Vector -> Statistic -> Vector Any
+    running data statistic=Statistic.Sum =
+        Statistic.running_bulk data [statistic] . map .first
+
+    ## Compute a set running statistics on a vector like object.
+
+       Arguments:
+       - data: Vector like object which has a `to_array` method.
+       - statistics: Set of statistics to calculate.
+    running_bulk : Vector -> Vector Statistic -> Vector Any
+    running_bulk data statistics=[Statistic.Count, Statistic.Sum] =
+        unsupported s = case s of
+            Statistic.Covariance _ -> True
+            Statistic.Pearson _ -> True
+            Statistic.Spearman _ -> True
+            Statistic.R_Squared _ -> True
+            _ -> False
+        if statistics.any unsupported then Error.throw (Illegal_Argument.Error ("Unsupported Statistics ( " + (statistics.filter unsupported . to_text) ") for running calculations.")) else
+            counter = CountMinMax.new
+
+            java_stats = statistics.map .order
+            skip_moments = java_stats.all s->s.is_nothing
+            moments = if skip_moments then Nothing else Moments.new (java_stats.filter v->v!=Nothing . fold 0 .max)
```

Comment:
Maybe I'm the only one confused by precedence here, but I'd add parentheses here to spare the reader from remembering the precedences of `->` and `!=`.

```suggestion
            moments = if skip_moments then Nothing else Moments.new (java_stats.filter (v-> v!=Nothing) . fold 0 .max)
```